app/main.py

```
@@ -96,10 +135,302 @@ def _tc_get_json(path, params=None, timeout=30):
+    except Exception as e:
+        logging.warning(f"Failed-step attribution failed for {btid}@{branch}: {e}; using config-level meta-runners")
+    return key, newest, attributed  # newest IS the current red build
+
+
+def update_failed_build_metrics(meta_runner_ids):
+    """Refresh FAILED_BUILD_GAUGE: one series per (candidate config, branch) that is
+    CURRENTLY failing -- i.e. failed within the window AND whose latest build on that branch
+    is still FAILURE (not yet recovered).
+    """
+    configs = enumerate_candidate_configs(meta_runner_ids)
+    logging.info(f"Candidate configs (use a monitored meta-runner): {len(configs)}")
```

Ruff (0.15.20)

[warning] Logging statement uses f-string

(G004)